frame/support/src/storage/counted_map.rs

```diff
@@ -0,0 +1,314 @@
+use codec::{FullCodec, Decode, EncodeLike, Encode};
+use crate::{
+	storage::{
+		StorageMap, StorageValue,
+		StorageAppend, StorageTryAppend, StorageDecodeLength, StoragePrefixedMap,
+		generator::{StorageMap as StorageMapT, StorageValue as StorageValueT},
+	},
+};
+use sp_std::prelude::*;
+use sp_runtime::traits::Saturating;
+
+/// A wrapper around a `StorageMap` and a `StorageValue<u32>` to keep track of how many items are in
+/// a map, without needing to iterate all the values.
+///
+/// This storage item has additional storage read and write overhead when manipulating values
+/// compared to a regular storage map.
+///
+/// For functions where we only add or remove a value, a single storage read is needed to check if
+/// that value already exists. For mutate functions, two storage reads are used to check if the
+/// value existed before and after the mutation.
+///
+/// Whenever the counter needs to be updated, an additional read and write occurs to update that
+/// counter.
+pub struct CountedStorageMap<Key, Value, Map, Counter>(
+	core::marker::PhantomData<(Key, Value, Map, Counter)>
+);
+
+impl<Key, Value, Map, Counter>
+	StorageMapT<Key, Value>
+	for CountedStorageMap<Key, Value, Map, Counter>
+where
+	Key: FullCodec,
+	Value: FullCodec,
+	Map: StorageMapT<Key, Value>,
+	Counter: StorageValueT<u32>
+{
+	type Query = Map::Query;
+	type Hasher = Map::Hasher;
+	fn module_prefix() -> &'static [u8] {
+		Map::module_prefix()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		Map::storage_prefix()
+	}
+	fn from_optional_value_to_query(v: Option<Value>) -> Self::Query {
+		Map::from_optional_value_to_query(v)
+	}
+	fn from_query_to_optional_value(v: Self::Query) -> Option<Value> {
+		Map::from_query_to_optional_value(v)
+	}
+}
+
+impl<Key, Value, Map, Counter>
+	StoragePrefixedMap<Value> for
+	CountedStorageMap<Key, Value, Map, Counter>
+where
+	Value: FullCodec,
+	Map: crate::StoragePrefixedMap<Value>,
+	Counter: StorageValueT<u32>
+{
+	fn module_prefix() -> &'static [u8] {
+		Map::module_prefix()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		Map::storage_prefix()
+	}
+}
+
+impl<Key, Value, Map, Counter>
+	CountedStorageMap<Key, Value, Map, Counter>
+where
+	Key: FullCodec,
+	Value: FullCodec,
+	Map: StorageMapT<Key, Value> + crate::StoragePrefixedMap<Value>,
+	Counter: StorageValueT<u32, Query=u32>,
+{
+	// Internal helper function to track the counter as a value is mutated.
+	fn mutate_counter<
+		KeyArg: EncodeLike<Key> + Clone,
+		M: FnOnce(KeyArg, F) -> R,
+		F: FnOnce(I) -> R,
+		I,
+		R,
+	>(m: M, key: KeyArg, f: F) -> R {
+		let val_existed = Map::contains_key(key.clone());
+		let res = m(key.clone(), f);
```

Comment:
```suggestion
		let res = m(&key, f);
```

Reply:
fixed, I use a reference now, I hope it doesn't get too much into reference of reference, this could be improved by having some inner function in frame-support storage. we could improve in order not to compute the final key multiple time.

Reply:
or maybe we should rely on the fact that the user gave the key in the most efficient way, so for big keys, it would be by reference and for small key like u32 by value.
In this case clone is more efficient than a reference.

Reply:
I think relying on user behavior is risky. Agree that if all users always do the right thing, then cloning here is more efficient than using a reference. However, I would expect that the performance impact of references-of-references to be very small; I would expect the impact of cloning a big key to be larger. Therefore, referencing seems the safer option.

Reply:
yes you are right, also my comment is wrong because user can pass a vec by value, without expected that it is cloned internally.

Reply:
If performance/efficiency is really an issue, making a trait bound on `key` and restricting it to be a `Copy` type would work as well, but that would mean that another function WITHOUT the `Copy` trait bound needs to be made, otherwise non-`Copy` types wouldn't be able to call this function.

Reply:
yes I guess we can refactor and bring some improvment this way.
But all in all it shouldn't be an issue, if we are to refactor I would prefer to focus on computing the final key only once, also maybe have some compile-time hash for storage_prefix and pallet_prefix.